### ISSUE TITLE
pypy alongside cpython + base/interpreter for noarch:python packages

### ISSUE THREAD

**Issue by mbargull** [msarahan/pypy-python-structure#1]
In reference to
@mattip in https://github.com/conda-forge/conda-forge.github.io/issues/867#issuecomment-555822040:
> Would a user ever get into a situation where they have two interpreters in the same conda environment?

Will Conda packages for CPython and PyPy themselves have path conflicts, e.g., both having `bin/python` or `lib/pythonX.Y/some-file`? IIRC, PyPy uses a different `lib/some-thing` path, right?

If there are no conflicts, we *could potentially* let them be installable side by side.

How would `conda` handle its `noarch: python` packages for PyPy? Currently, it uses `python.exe` or `bin/pythonX.Y`. Either the `pypy` package, or the `python=*=pypy` package should provide those binaries. If we put them into the `python` package, we could do the same for `cpython` and ensure exclusivity. (I guess this may put off some people since `cpython` wouldn't be considered "complete" then. Well, I'm just putting down thoughts as them come right now and still lack a view of the *whole* picture...)

**Comment by mattip** [comment]
> IIRC, PyPy uses a different lib/some-thing path, right?

That is a decision for conda to make. As packaged in the PyPy-provided tarballs, PyPy's stdlib and site-packages paths are different, but not for any good reason. Typically packages like debian, arch modify this as they wish. If conda wishes to put the two iplementaitons into the same environment they should ensure the layouts are different, and that the typical symlink of `bin/pypy` -> `bin/python` is not created.
I would advise against it as "too complicated" but I don't know how sensitive users are to disk space

**Comment by mattip** [comment]
Would this design be valid for building feedstocks that use c-extension modules like NumPy and lxml? They cannot be shared across implementations.

**Comment by msarahan** [comment]
yes.  Those packages would use the python-devel package shown in the recipe here.  That would add a runtime dependency on either pypy or cpython to prevent ambiguity and problems.  It might need tweaking beyond the example here, but the rough idea here should work.

**Comment by mbargull** [comment]
> \[...\] PyPy's stdlib and site-packages paths are different, but not for any good reason. Typically packages like debian, arch modify this as they wish.

Good to know, thanks!

> \[...\] I would advise against it as "too complicated"

:+1: TBH, that was what I anticipated to hear. IMHO, there is not too much value in having CPython and PyPy in the *same* environment.
(I mean, conda-forge and Anaconda don't even support running a Python 2 interpreter in the same environment as a Version 3 one, i.e., we don't offer separate `python2` packages -- so not offering CPython and PyPy in the same env should not come as too much of a surprise for the user.)
